File: solutions/Medium/Generate_Parentheses_0022.py
from typing import List


class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        """
        Function: generateParenthesis
        Description: Generates all valid combinations of n pairs of parentheses.

        Parameters:
        - n (int): Number of pairs of parentheses.

        Returns:
        - List[str]: A list of all well-formed parentheses combinations.
        """
        def backtrack(s, open_count, close_count):
            # Base Case: Valid combination found
            if len(s) == 2 * n:
                result.append(s)
                return

            # Add an open parenthesis if we haven't reached the limit
            if open_count < n:
                backtrack(s + "(", open_count + 1, close_count)

            # Add a closing parenthesis if it does not exceed open count
            if close_count < open_count:
                backtrack(s + ")", open_count, close_count + 1)

        result = []
        backtrack("", 0, 0)

        return result


# Example Test Cases
if __name__ == "__main__":
    solution = Solution()

    # Test Case 1: Smallest case with n = 1
    print("Test Case 1: n = 1")
    print("Expected Output: ['()'] | Actual Output:",
          solution.generateParenthesis(1))

    # Test Case 2: Standard case with n = 2
    print("Test Case 2: n = 2")
    print("Expected Output: ['(())', '()()'] | Actual Output:",
          solution.generateParenthesis(2))

    # Test Case 3: n = 3 (More complex case)
    print("Test Case 3: n = 3")
    print("Expected Output: ['((()))', '(()())', '(())()', '()(())', '()()()'] | Actual Output:",
          solution.generateParenthesis(3))

    # Test Case 4: Edge case with n = 4
    print("Test Case 4: n = 4")
    print("Expected Output: 14 unique combinations | Actual Output:",
          len(solution.generateParenthesis(4)))

    # Test Case 5: Larger input (n = 5)
    print("Test Case 5: n = 5")
    print("Expected Output: 42 unique combinations | Actual Output:",
          len(solution.generateParenthesis(5)))

    # Test Case 6: Maximum constraint (n = 8)
    print("Test Case 6: n = 8")
    print("Expected Output: A large number of valid combinations | Actual Output:", len(
        solution.generateParenthesis(8)))

"""
Time Complexity:
- O(4^n / sqrt(n)) → Catalan number growth rate for valid sequences.

Space Complexity:
- O(4^n / sqrt(n)) → To store all valid sequences.
"""

File: solutions/Medium/Generate_Parentheses_0022.js
/**
 * Function: generateParenthesis
 * Description: Given `n` pairs of parentheses, this function generates all possible 
 * well-formed combinations using backtracking.
 *
 * @param {number} n - Number of pairs of parentheses.
 * @returns {string[]} - An array containing all valid combinations of parentheses.
 */
const generateParenthesis = function(n) {
    let result = [];

    /**
     * Backtracking function to generate valid parentheses combinations.
     * @param {string} combination - The current combination being formed.
     * @param {number} open - Number of open parentheses used.
     * @param {number} close - Number of closed parentheses used.
     */
    const backtrack = (combination, open, close) => {
        // If the combination reaches the required length, add it to the result
        if (combination.length === 2 * n) {
            result.push(combination);
            return;
        }

        // Add an open parenthesis if it does not exceed `n`
        if (open < n) {
            backtrack(combination + "(", open + 1, close);
        }

        // Add a close parenthesis if it does not exceed the open count
        if (close < open) {
            backtrack(combination + ")", open, close + 1);
        }
    };

    // Start the backtracking process with an empty string
    backtrack("", 0, 0);

    return result;
};

// Example Test Cases
if (require.main === module) {
    console.log("Test 1: Basic case with n = 3");
    console.log(generateParenthesis(3));
    // Expected output: ["((()))","(()())","(())()","()(())","()()()"]

    console.log("Test 2: Smallest case with n = 1");
    console.log(generateParenthesis(1));
    // Expected output: ["()"]

    console.log("Test 3: Case with n = 2");
    console.log(generateParenthesis(2));
    // Expected output: ["(())", "()()"]

    console.log("Test 4: Edge case with maximum n = 8 (should generate 1430 valid combinations)");
    console.log(generateParenthesis(8).length);
    // Expected output: 1430

    console.log("Test 5: Case with n = 4");
    console.log(generateParenthesis(4));
    // Expected output: ["(((())))","((()()))","((())())","((()))()","(()(()))","(()()())","(()())()","(())(())","(())()()","()((()))","()(()())","()(())()","()()(())","()()()()"]

    console.log("Test 6: Case with n = 5");
    console.log(generateParenthesis(5).length);
    // Expected output: 42 (Number of valid combinations for n = 5)
}

//  Complexity analysis
// Time Complexity: O(4^n / sqrt(n)) - This is the Catalan number growth rate.
// Space Complexity: O(4^n / sqrt(n)) - Due to storing all valid combinations in the result array.

export default generateParenthesis;